packages/neon-api/src/transaction/validator.ts

```diff
@@ -0,0 +1,298 @@
+   * Whether the transaction is valid after validation
+   */
+  valid: boolean;
+  suggestions?: {
+    validaUntilBlock?: ValidationSuggestion<number>;
+    script?: Array<string>;
+    systemFee?: ValidationSuggestion<u.Fixed8>;
+    networkFee?: ValidationSuggestion<u.Fixed8>;
+  };
+}
+
+/**
+ * A class with functions to validate transaction
+ */
+export class TransactionValidator {
+  /**
+   * Transaction will be validated on this rpc node
+   */
+  public rpcClient: rpc.RPCClient;
+
+  /**
+   * Transaction that will be validated
+   */
+  public transaction: tx.Transaction;
+
+  constructor(rpc: rpc.RPCClient, transaction: tx.Transaction) {
+    this.rpcClient = rpc;
+    this.transaction = transaction;
+  }
+
+  /**
+   * validate transaction attribute - validUntilBlock
+   * @param autoFix will automatically fix transaction if specified as true
+   */
+  async validateValidUntilBlock(autoFix = false): Promise<ValidationResult> {
+    const { validUntilBlock } = this.transaction;
+    const height = await this.rpcClient.getBlockCount();
+    if (
+      validUntilBlock <= height ||
+      validUntilBlock >= height + tx.Transaction.MAX_TRANSACTION_LIFESPAN
+    ) {
+      const vub_suggestion =
+        tx.Transaction.MAX_TRANSACTION_LIFESPAN + height - 1;
+      if (autoFix) {
+        this.transaction.validUntilBlock = vub_suggestion;
+        return {
+          valid: true,
+          suggestions: {
+            validaUntilBlock: {
+              fixed: true,
+              prev: validUntilBlock,
+              suggestion: vub_suggestion
+            }
+          }
+        };
+      }
+      return {
+        valid: false,
+        suggestions: {
+          validaUntilBlock: {
+            fixed: false,
+            prev: validUntilBlock,
+            suggestion: vub_suggestion
+          }
+        }
+      };
+    }
+    return {
+      valid: true
+    };
+  }
+
+  private async _validateIntent(
+    intent: sc.ScriptIntent
+  ): Promise<string | undefined> {
+    const { scriptHash, operation } = intent;
+    const manifest = await this.rpcClient.getContractState(scriptHash);
+    if (manifest === null) {
+      return `Unknow contract ${scriptHash}`;
+    }
+
+    if (operation) {
+      if (
+        manifest.abi.methods.map(method => method.name).indexOf(operation) < 0
+      ) {
+        return `Unknow method ${operation} for constract ${scriptHash}`;
+      }
+    }
+
+    return undefined;
+  }
+
+  /**
+   * Validate intents
+   */
+  async validateScript(): Promise<ValidationResult> {
+    const intents = new sc.ScriptParser(
+      this.transaction.script
+    ).toScriptParams();
+    const intentsRes: Array<string> = [];
+    await Promise.all(
+      intents.map(intent =>
+        this._validateIntent(intent).then(res =>
+          res ? intentsRes.push(res) : res
+        )
+      )
+    );
+    if (intentsRes.length > 0) {
+      return {
+        valid: false,
+        suggestions: {
+          script: intentsRes
```

Comment:
We should return a ValidationSuggestion and add a `message` field that indicates a human readable error. At this point, you might want to not add the `message` feature first and complete this PR.

Reply:
I just added this. Please check